### PR TITLE
Stop ignoring the mempool conflicting transaction reject list size limit

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

If `MempoolStorage::insert` gets a transaction spend conflict error, it returns that error early.
This early return skips the mempool reject list limits, allowing those lists to have an unlimited size.

This is a memory denial of service security issue.
It is not in any release - the reject list limits were only added in PR #2844.

(The small mempool size does not prevent this issue, because a small mempool can still have conflicts.)

## Solution

- always reject using `MempoolStorage::reject`, which enforces the size limits for every reject list
- add rejection list size tests for the two errors returned by `MempoolStorage::insert`
- add a general rejection list size test for randomly generated rejection errors

## Review

Anyone can review this PR.

It's based on PR #2852 and the latest `main` branch.

@conradoplg might want to review commit 51e59aec7402204fa507ac55fb1fb9993b2f8456,
which resolves a tricky conflict between his PR #2852 and PR #2826.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebrad/src/components/mempool/storage.rs [hunk 1]
```diff
@@ -136,33 +143,36 @@ impl Storage {
             return Err(MempoolError::InMempool);
         }
 
-        // Then, we try to insert into the pool. If this fails the transaction is rejected.
+        // Try to insert into the pool. If this fails the transaction is rejected.
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Not needed comment change
```suggestion
        // Then, we try to insert into the pool. If this fails the transaction is rejected.
```

File: zebrad/src/components/mempool/storage.rs [hunk 2]
```diff
@@ -137,32 +144,35 @@ impl Storage {
...
-                .entry(SameEffectsChainRejectionError::RandomlyEvicted)
-                .or_default()
-                .insert(evicted_tx.id.mined_id());
+            self.reject(
+                evicted_tx.id,
+                SameEffectsChainRejectionError::RandomlyEvicted.into(),
+            );
+
+            // If this transaction gets evicted, set its result to the same error
+            // (we could return here, but we still want to check the size)
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]:
```suggestion
            // (we could return here, but we still want to check the mempool size)
```

**Comment by conradoplg** [comment]
I reviewed this and looks good, but I'll wait for https://github.com/ZcashFoundation/zebra/pull/2852 to merge before merging this

**Comment by teor2345** [comment]
> I reviewed this and looks good, but I'll wait for #2852 to merge before merging this

Yeah it might need a rebase after #2852 merges.

**Comment by teor2345** [comment]
I'm just going to admin-merge this, because the previous CI passed, and the latest PR is unrelated to this code.

**Pull request merged**